src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -2842,15 +2838,55 @@ impl Coordinator {
                 return Err(AdapterError::Unsupported("EXPLAIN TIMESTAMP AS DOT"));
             }
         };
-        let timeline_context = self.validate_timeline_context(optimized_plan.depends_on())?;
-        let determination = self.determine_timestamp(
+        let conn_id = session.conn_id();
+        let source_ids = source.depends_on();
+        let timeline_context = self.validate_timeline_context(source_ids.clone())?;
+
+        let mut determination = self.determine_timestamp(
             session,
             &id_bundle,
             &QueryWhen::Immediately,
             cluster_id,
-            timeline_context,
+            timeline_context.clone(),
             real_time_recency_ts,
         )?;
+
+        if session.transaction().is_in_multi_statement_transaction() {
+            if let Some(ts_context @ TimestampContext::TimelineTimestamp(_, _)) =
+                session.get_transaction_timestamp_context()
+            {
+                // In this case, `timestamp_context` should be updated to reflect the one
+                // chosen as a result of the transaction's initial peek.
+                determination.timestamp_context = ts_context;
+            } else {
+                if matches!(session.transaction(), &TransactionStatus::InTransaction(_)) {
+                    session.add_transaction_ops(TransactionOps::Peeks(
+                        determination.timestamp_context.clone(),
+                    ))?;
+                }
+            }
+
+            self.check_txn_timedomain_conflicts(
+                session,
+                &determination.timestamp_context,
+                &id_bundle,
+            )?;
+
+            // If we're in a transaction, acquire read holds on any sources in the current time-domain
+            // if they have not already been acquired.
+            let timedomain_id_bundle =
+                self.timedomain_for(&source_ids, &timeline_context, conn_id, cluster_id)?;
+
+            if !timedomain_id_bundle.is_empty() && !self.txn_reads.contains_key(&session.conn_id())
+            {
+                if let Some(timestamp) = determination.timestamp_context.timestamp() {
+                    let read_holds =
+                        self.acquire_read_holds(timestamp.clone(), &timedomain_id_bundle);
+                    self.txn_reads.insert(session.conn_id(), read_holds);
+                }
+            }
+        }
+
```

Comment:
All of this logic makes sense to me, I'm assuming you got it from the sequence peek methods? I'm slightly concerned about duplicating this logic between peeks and explain timestamps. The logic is a bit nuanced and it will be easy for it to accidentally diverge in the future. Is there some way to avoid duplicating this?

Reply:
Definitely, was thinking that as well. I think both paths share pretty much all of this logic so I'll try factoring it out

Reply:
@jkosh44 Did a couple things:

1. Lifted the logic to use transactional timestamp context to `determine_timestamp`. Feels like this is an appropriate place for it, but let me know if I’m missing something.
2. Factored out common logic around determining a query’s timestamp, acquiring read holds, and tracking txn ops into `sequence_peek_timestamp`. I couldn’t quite see reusing `sequence_peek_stage` as the planning stage deals with dataflow-specific logic that `EXPLAIN TIMESTAMP` doesn’t depend on.

Appreciate any and all feedback!